cmd/rukpakctl/cmd/run.go

```diff
@@ -0,0 +1,187 @@
+			if err != nil {
+				log.Fatalf("failed to upload bundle: %v", err)
+			}
+			if !modified {
+				fmt.Printf("bundle %q is already up-to-date\n", bundleName)
+				return
+			}
+
+			fmt.Printf("successfully uploaded bundle content for %q\n", bundleName)
+		},
+	}
+	cmd.Flags().StringVar(&systemNamespace, "system-namespace", "rukpak-system", "the namespace in which the rukpak controllers are deployed.")
+	cmd.Flags().StringVar(&uploadServiceName, "upload-service-name", "upload-manager", "the name of the service of the upload manager.")
+	cmd.Flags().StringVar(&caSecretName, "ca-secret-name", "rukpak-ca", "the name of the secret in the system namespace containing the root CAs used to authenticate the upload service.")
+	cmd.Flags().StringVar(&bundleDeploymentProvisionerClassName, "bundle-deployment-provisioner-class", "core.rukpak.io/plain", "Provisioner class name to set on bundle deployment.")
+	cmd.Flags().StringVar(&bundleProvisionerClassName, "bundle-provisioner-class", "core.rukpak.io/plain", "Provisioner class name to set on bundle.")
+	return cmd
+}
+
+func buildBundleDeployment(bdName string, bundleLabels map[string]string, biPCN, bPNC string) *unstructured.Unstructured {
+	// We use unstructured here to avoid problems of serializing default values when sending patches to the apiserver.
+	// If you use a typed object, any default values from that struct get serialized into the JSON patch, which could
+	// cause unrelated fields to be patched back to the default value even though that isn't the intention. Using an
+	// unstructured ensures that the patch contains only what is specified. Using unstructured like this is basically
+	// identical to "kubectl apply -f"
+	return &unstructured.Unstructured{Object: map[string]interface{}{
+		"apiVersion": rukpakv1alpha1.GroupVersion.String(),
+		"kind":       rukpakv1alpha1.BundleDeploymentKind,
+		"metadata": map[string]interface{}{
+			"name": bdName,
+		},
+		"spec": map[string]interface{}{
+			"provisionerClassName": biPCN,
+			"template": map[string]interface{}{
+				"metadata": map[string]interface{}{
+					"labels": bundleLabels,
+				},
+				"spec": map[string]interface{}{
+					"provisionerClassName": bPNC,
+					"source": map[string]interface{}{
+						"type":   rukpakv1alpha1.SourceTypeUpload,
+						"upload": &rukpakv1alpha1.UploadSource{},
+					},
+				},
+			},
+		},
+	}}
+}
+
+func getBundleName(ctx context.Context, cfg *rest.Config, bundleLabels map[string]string) (string, error) {
+	dynCl, err := dynamic.NewForConfig(cfg)
+	if err != nil {
+		return "", fmt.Errorf("build dynamic client: %v", err)
+	}
+
+	watch, err := dynCl.Resource(rukpakv1alpha1.GroupVersion.WithResource("bundles")).Watch(ctx, metav1.ListOptions{Watch: true, LabelSelector: labels.FormatLabels(bundleLabels)})
+	if err != nil {
+		return "", fmt.Errorf("watch bundles: %v", err)
+	}
+	defer watch.Stop()
+
+	select {
+	case evt := <-watch.ResultChan():
+		return evt.Object.(client.Object).GetName(), nil
```

Comment:
Do you think we need to type assert here to avoid a panic? I feel like this should be largely fine.

Reply:
If this starts panic-ing, something pretty drastic would have changed, I think we'd have more serious issues. Essentially, the bet we're making is that the event Object contains something that also implements `metav1.Object`, which seems like a safe bet until at least k8s 2.0.

Reply:
Works for me :+1: